### PR TITLE
Feat: Ignore differences owned by trusted managers from managedFields

### LINKED ISSUE #2913: Maintain difference in cluster and git values for specific fields
# Summary
If another controller modifies a field in a resource and that field is required by the resource to be created, then that resource cannot be compatible with GitOps. The manifest stored in Git requires that field in order to be created, but another controller will modify that field once created making it out of sync. Argo CD should have an annotation called `argocd.argoproj.io/respect-existing-fields` that accepts JSON paths to fields that are managed by other controllers. That annotation should use stored in the Kubernetes instead of applying what is in Git when Argo CD does a diff or apply. This is a step further than https://argoproj.github.io/argo-cd/user-guide/diffing/#application-level-configuration

# Motivation
Istio Virtual Services require a weight when two destinations are listed in a route. If a user creates the Virtual Service using GitOps and another controller modifies those weights to shape traffic, the Virtual Service will be out of sync with what is stored in Git. If there is a Sync, the Virtual Service will be revert to the Git state and then the other controller should revert it back.

# Proposal

How do you think this should be implemented?

The application controller should look for this annotation on all objects within the repo while executing a diff or sync. If found, the diffing logic should not figure this information into Diff status, and the apply logic should not change this value in the cluster.

Questions/Concerns that should be addressed during the implementation:
- How should the annotation be applied if the annotation itself is modified in Git?
- The `argocd app get manifest` command should return the actual Git contents instead of the modified Git contents.
- The annotation should be able to index an entire array (i.e. `spec.http.[*].route[*].weight` means check each entry of http and the entries within each route.)
- Allow multiple fields to ignore

### PR DESCRIPTION
**Pull request by leoluz**
Add new config in `Application.ignoreDifferences` to allow configuring managers that should be ignored during diffs between live and desired states.

Partially Fix https://github.com/argoproj/argo-cd/issues/2913

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

File: controller/state.go [hunk 1]
```diff
@@ -505,6 +508,18 @@ func (m *appStateManager) CompareAppState(app *v1alpha1.Application, project *ap
 
 	for i := range reconciliation.Target {
 		_ = m.resourceTracking.Normalize(reconciliation.Target[i], reconciliation.Live[i], appLabelKey, string(trackingMethod))
+		// just normalize on managed fields if live and target aren't nil as we just care
+		// about conflicting fields
+		if reconciliation.Live[i] != nil && reconciliation.Target[i] != nil {
+			t := reconciliation.Target[i]
+			gvk := t.GetObjectKind().GroupVersionKind()
+			if ok, ignoreDiff := ignoreDiffConfig.HasIgnoreDifference(gvk.Group, gvk.Kind, t.GetName(), t.GetNamespace()); ok {
```
Comment (mayzhang2000) [inline controller/state.go:516]: Is this checking of `ignoreDiffConfig.HasIgnoreDifference` needed?

File: docs/user-guide/diffing.md [hunk 2]
```diff
@@ -53,6 +55,18 @@ spec:
...
+```yaml
+spec:
+  ignoreDifferences:
+  - group: *
+    kind: *
+    managedFieldsManagers:
+    - kube-controller-manager
+```
+
+The above configuration will ignore differences from all fields owned by `kube-controller-manager` for all resources belonging to this application.
```
Comment (alexmt) [inline docs/user-guide/diffing.md:68]: Please add example of `managedFieldsManagers` for system-level configuration.

Also please add `managedFieldsManagers` to https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/application.yaml and https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/argocd-cm.yaml .

Comments might be minimal. These two files are used to quickly find which settings are available.

**Review by alexmt — CHANGES_REQUESTED** [review]
Thank you @leoluz ! Add only few minor comments. LGTM after resolving

File: controller/state.go [hunk 3]
```diff
@@ -651,6 +662,67 @@ func (m *appStateManager) CompareAppState(app *v1alpha1.Application, project *ap
...
+// preDiffNormalize applies the normalization of live and target resources before invoking
+// the diff. None of the attributes in the preDiffNormalizeParams will be modified. The
+// normalizeResults will return a list of ApplicationConditions in case something goes
+// wrong during the normalization.
+func (m *appStateManager) preDiffNormalize(p *preDiffNormalizeParams) *normalizeResults {
+	results := &normalizeResults{}
+	for i := range p.targets {
+		target := safeDeepCopy(p.targets[i])
+		live := safeDeepCopy(p.lives[i])
+		_ = m.resourceTracking.Normalize(target, live, p.appLabel, p.trackingMethod)
```
Comment (mayzhang2000) [inline controller/state.go:688]: I am not sure if the input of `resourceTracking.Normalize` should be the original value or the copy.

File: Procfile [hunk 4]
```diff
@@ -1,8 +1,11 @@
...
 redis: bash -c "if [ \"$ARGOCD_REDIS_LOCAL\" == 'true' ]; then redis-server --save '' --appendonly no --port ${ARGOCD_E2E_REDIS_PORT:-6379}; else docker run --rm --name argocd-redis -i -p ${ARGOCD_E2E_REDIS_PORT:-6379}:${ARGOCD_E2E_REDIS_PORT:-6379} redis:6.2.4-alpine --save '' --appendonly no --port ${ARGOCD_E2E_REDIS_PORT:-6379}; fi"
 repo-server: sh -c "FORCE_LOG_COLORS=1 ARGOCD_FAKE_IN_CLUSTER=true ARGOCD_GNUPGHOME=${ARGOCD_GNUPGHOME:-/tmp/argocd-local/gpg/keys} ARGOCD_PLUGINSOCKFILEPATH=${ARGOCD_PLUGINSOCKFILEPATH:-/tmp/argo-e2e/app/config/plugin}  ARGOCD_GPG_DATA_PATH=${ARGOCD_GPG_DATA_PATH:-/tmp/argocd-local/gpg/source} ARGOCD_TLS_DATA_PATH=${ARGOCD_TLS_DATA_PATH:-/tmp/argocd-local/tls} ARGOCD_SSH_DATA_PATH=${ARGOCD_SSH_DATA_PATH:-/tmp/argocd-local/ssh} ARGOCD_BINARY_NAME=argocd-repo-server ARGOCD_GPG_ENABLED=${ARGOCD_GPG_ENABLED:-false} go run ./cmd/main.go --loglevel debug --port ${ARGOCD_E2E_REPOSERVER_PORT:-8081} --redis localhost:${ARGOCD_E2E_REDIS_PORT:-6379}"
```
[the remark this answers is not in the document]
Comment (leoluz) [reply Procfile:7]: Pretty neat! great suggestion. Tks!
BTW, I introduced the bin usage in goreman to be able to attach a debugging session to it. It seems `go run` processes don't export the symbols necessary for debugging.

File: controller/state.go [hunk 5]
```diff
@@ -651,6 +662,67 @@ func (m *appStateManager) CompareAppState(app *v1alpha1.Application, project *ap
...
+// safeDeepCopy will return nil if given obj is nil.
+func safeDeepCopy(obj *unstructured.Unstructured) *unstructured.Unstructured {
+	if obj == nil {
+		return nil
+	}
+	return obj.DeepCopy()
+}
+
+func appCondition(t v1alpha1.ApplicationConditionType, err error) v1alpha1.ApplicationCondition {
+	now := metav1.Now()
```
[the remark this answers is not in the document]
Comment (leoluz) [reply controller/state.go:718]: done

File: Procfile (hunk 4 above)
[the remark this answers is not in the document]
Comment (leoluz) [reply Procfile:7]: done

File: controller/state.go (hunk 3 above)
Comment (leoluz) [reply controller/state.go:688]: As we discussed, this change also fix another bug where the code was mutating the target resource and it was reflected in the sync when it shouldn't. The new code will never mutate the objects that are used during the sync

File: docs/user-guide/diffing.md (hunk 2 above)
Comment (leoluz) [reply docs/user-guide/diffing.md:68]: done

**Review by alexmt — APPROVED** [review]
LGTM

**Pull request merged**